add pip configuration

Meet-synaptics opened the pull request.

Meet-synaptics committed:
add pip configuration

Meet-synaptics committed:
Merge remote-tracking branch 'origin/main' into wip/add-pip-setup

Meet-synaptics committed:
modify workflow to trigger on release

lkeller-synaptics commented on .github/workflows/build-release-pip-package.yml:
```diff
@@ -1,12 +1,8 @@
-        required: true
-        default: 'main'
+  release:
+    types: [published, prereleased]
```
AFAIK, this will trigger twice for pre-releases if we first created a draft pre-release and then publish it. I think "prereleased" event is not necessary?

Maybe the best hook would be that we trigger when a draft release is created so that when we publish it, it already has the asset. This would mean to trigger on "created".

Meet-synaptics replied:
As per the documentation, the prereleased type will not trigger for pre-releases published from draft releases, but the published type will trigger. It will only trigger once.
https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#release

Yes, that's the best way. When we create the draft release, the workflow should attach the Python wheel to it so we can test it from the draft release itself. Unfortunately, workflows can't be triggered for the created, edited, or deleted activity types for draft releases.

One option is to run a manual workflow once the draft release is created.?

lkeller-synaptics commented on .github/workflows/build-release-pip-package.yml:
```diff
@@ -21,7 +17,7 @@ jobs:
       uses: actions/checkout@v4
       with:
-        ref: ${{ github.event.inputs.branch }}
+        ref: main
```
Here we should checkout the code that will be tagged in the branch (I think it is event.release.target_commitish).

lkeller-synaptics commented on .github/workflows/build-release-pip-package.yml:
```diff
@@ -53,65 +49,19 @@ jobs:
           pip install --upgrade twine
 
-    - name: Run build and deploy github release
+    - name: Run build
```
Here it would be nice to set an env to provide the name of the version to the setup script.

One way would look something like this:

env:
   VERSION: ${{ event.release.prerelease && "snapshot" || event.release.tag_name}}

lkeller-synaptics commented on setup.py:
```diff
@@ -0,0 +1,64 @@
+from datetime import datetime
+
+# Get the current date and time as version
+def get_version():
```
if in the workflow we add the VERSION env variable here we can do something like this:

```
if os.environ.get('VERSION', 'snapshot') == 'snapshot':
    return 'snapshot-' + datetime.now().strftime("%Y%m%d%H%M%S")
else:
    return os.environ['VERSION']
```

Meet-synaptics replied:
The standard naming convention for a Python package is package_name-version-distribution.whl. For example, synap-3.1.0-cp310-cp310-linux_x86_64.whl.

In our case, name of wheel will be like "synap-snapshot-20240625200405-cp310-cp310-linux_x86_64.whl"
,existing datetime.now().strftime("%Y%m%d%H%M%S") approach is not suitable for pip naming conventions because it generates timestamps that do not provide meaningful versioning information required by package managers like pip. if possible, could we adopt a method to derive proper versioning from toolkit or maybe pysynap for better clarity and consistency?

lkeller-synaptics replied:
as discussed over the phone I agree what I proposed above is not good.

tensorflow uses the schema X.Y.ZrcW (see https://pypi.org/project/tensorflow/#history), we could do something similar like:

3.1.0snapshotYYMMDD where the first version part comes from the hardcoded version number in the source code as you were doing before.

Meet-synaptics replied:
That make sense to use "X.YrcN  # Release Candidate" format as per official standards: https://peps.python.org/pep-0440/

It would be look like this;
synap-3.1.0rc20240703222235-cp310-cp310-linux_x86_64.whl

lkeller-synaptics commented on synap_convert.py:
```diff
@@ -87,6 +87,9 @@ def main():
         # Assume the directory layout of the public toolkit directory
         synap_dir = pathlib.Path(__file__).parent.parent.absolute()
         tools_dir = os.path.join(synap_dir, "toolkit-prebuilts")
+        # check if prebuilts is avaiblle alongside
```
s/avaiblle/available/

mdenna-synaptics reviewed (commented):
I think the week name should also contain the version (eg 3.2.0) and the supported py interpreter, something like:

synap-3.2.0-cp310-cp310-linux_x86_64-py3.10-20240625200405.whl

?
I know it's long but at least we have all the info specified. We could drop the final build date and number in the official release.

ayounes-synaptics commented:
We need to support both Python 3.10 (Ubuntu 22.04) and 3.12 (Ubuntu 24.04). In order to do this we need to prepare toolkit-prebuilt accordingly, especially the pyarmor part:

https://pyarmor.readthedocs.io/en/latest/tutorial/advanced.html#obfuscating-scripts-for-multiple-python-versions

ayounes-synaptics commented:
About toolkit this the versio-less requirements.txt that I use. This is triggering all 2nd level dependencies automatically:
 
tensorflow-cpu
onnxoptimizer
onnxruntime
onnxsim
dill
ruamel.yaml
onnx-tf
ply
future
torch
torchvision
torchaudio

I understand and agree that we need to set versions in there but Ideally the pip wheel package should allow ranges of tested versions. For example tensorflow-cpu between 2.13 and 2.16.

ayounes-synaptics commented:
tensorflow>=2.13.0+cpu
onnx>=1.14.0
onnxoptimizer>=0.3.13
onnxruntime>=1.15.1
onnxsim>=0.4.33
dill>=0.3.7
ruamel.yaml>=0.17.32
onnx-tf>=1.10.0
ply>=3.11
future>=0.18.3
torch>=1.13.1+cpu
torchvision>=0.14.1+cpu
torchaudio>=0.13.1+cpu

Meet-synaptics commented:
> tensorflow>=2.13.0+cpu onnx>=1.14.0 onnxoptimizer>=0.3.13 onnxruntime>=1.15.1 onnxsim>=0.4.33 dill>=0.3.7 ruamel.yaml>=0.17.32 onnx-tf>=1.10.0 ply>=3.11 future>=0.18.3 torch>=1.13.1+cpu torchvision>=0.14.1+cpu torchaudio>=0.13.1+cpu


It's not possible as local version label (+cpu) can only be used with == or != operators.

Meet-synaptics committed:
-Simplified requiremnet.txt
-Update wheel package name as per PEP440 Naming standards
-Update install_require for wheel package

Meet-synaptics committed:
-Simplified requiremnet.txt
-Update wheel package name as per PEP440 Naming standards
-Update install_require for wheel package
-Update github action to upload wheels in specified draft release

Meet-synaptics committed:
Merge branch 'wip/add-pip-setup' of github.com:synaptics-synap/toolkit into wip/add-pip-setup

Modify workflow to upload wheel in same job

Meet-synaptics committed:
Update workflow

-remove deprecation warning for Node.js
-build only binary distribution

Meet-synaptics committed:
Merge branch 'main' of github.com:synaptics-synap/toolkit into wip/add-pip-setup

Meet-synaptics committed:
- Added image_od, image_from_raw, and image_to_raw tools to the wheel
- Excluded unnecessary files from the wheel
- Incremented toolkit version
- Added console script for the 'help' command

lkeller-synaptics commented on setup.py:
```diff
@@ -40,21 +40,40 @@ def get_requirements(directory):
-long_description = (this_directory / "README.rst").read_text()
+# Create synap_help.py script for handling help command
+with open('pysynap/scripts/synap_help.py', 'w') as f:
+    f.write("""\
```
Maybe it would be better to have this also as a file like the other scripts or the idea is to later auto-generate it based on the scripts we install (so that we don't need to maintain the help in two places)?

Meet-synaptics replied:
I think yes It's better to have a file rather than to create it dynamically.

mdenna-synaptics reviewed (approved):
looks good to me

Meet-synaptics committed:
Add help console script

-Remove dynamic creation of synap_help.py
